File: Android/Android_Component/Components/app/src/main/java/ramon/better/com/components/activity/ServiceTestActivity.java
package ramon.better.com.components.activity;

import android.app.Service;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import ramon.better.com.components.R;
import ramon.better.com.components.service.BindService;
import ramon.better.com.components.service.StartService;

public class ServiceTestActivity extends AppCompatActivity {
    private static final String TAG = "BindServiceTag";
    // start 的方式
    private Button startServiceBtn;
    private Button stopServiceBtn;
    // bind 的方式
    private Button bindServiceBtn;
    private Button unbindServiceBtn;
    private Button getStatusBtn;
    private Context mContext;


    //保持所启动Service的IBinder对象
    BindService.MyBinder binder;



    //定义一个ServiceConnection对象
    private ServiceConnection connection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
            Log.i(TAG,"service connected");
            //获取Service的onBind方法所返回的MyBinder对象
            binder = (BindService.MyBinder) iBinder;
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            Log.i(TAG,"Service disconnected");
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_service_test);
        init();
    }

    private void init() {
        mContext = ServiceTestActivity.this;
        startServiceBtn = (Button)findViewById(R.id.start_service_btn);
        startServiceBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(mContext, StartService.class);
                mContext.startService(intent);
            }
        });
        stopServiceBtn = findViewById(R.id.stop_service_btn);
        stopServiceBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(mContext, StartService.class);
                mContext.stopService(intent);
            }
        });
        bindServiceBtn = (Button)findViewById(R.id.bind_service_btn);
        bindServiceBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // 绑定服务
                Intent bindIntent = new Intent(mContext, BindService.class);
                bindService(bindIntent, connection, Service.BIND_AUTO_CREATE);
            }
        });

        unbindServiceBtn = (Button)findViewById(R.id.unbind_service_btn);
        unbindServiceBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // 解除绑定
                unbindService(connection);
            }
        });
        // 获取 service 状态
        getStatusBtn = (Button)findViewById(R.id.get_status_btn);
        getStatusBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //获取显示Servcie的count值
                Toast.makeText(mContext, "count value is " + binder.getCount(),Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: Android/FourComponent/app/src/main/java/ramon/lee/fourcomponent/activity/ServiceTestActivity.kt
package ramon.lee.fourcomponent.activity

import android.app.Service
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.databinding.DataBindingUtil
import ramon.lee.fourcomponent.R
import ramon.lee.fourcomponent.databinding.ActivityServiceTestBinding
import ramon.lee.fourcomponent.service.BindService
import ramon.lee.fourcomponent.service.BindService.MyBinder
import ramon.lee.fourcomponent.service.MyIntentService
import ramon.lee.fourcomponent.service.StartService


class ServiceTestActivity : AppCompatActivity() {
    companion object {
        private const val TAG = "ServiceTestActivity"
    }
    private var binding: ActivityServiceTestBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_service_test)
        initView()
    }

    //保持所启动Service的IBinder对象
    var binder: MyBinder? = null
    //定义一个ServiceConnection对象
    private val connection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
            Log.i(TAG, "service connected")
            //获取Service的onBind方法所返回的MyBinder对象
            binder = iBinder as MyBinder
        }

        override fun onServiceDisconnected(componentName: ComponentName) {
            Log.i(TAG, "Service disconnected")
        }
    }


    private fun initView() {
        binding?.let {
            it.btnStartService.setOnClickListener {
                val intent = Intent(this, StartService::class.java)
                startService(intent)

                // 为了触发 onRebind ，startService 启动 BindService
                val bindIntent = Intent(this, BindService::class.java)
                startService(bindIntent)
            }

            it.btnStopService.setOnClickListener {
                val intent = Intent(this, StartService::class.java)
                stopService(intent)

                val bindIntent = Intent(this, BindService::class.java)
                stopService(bindIntent)
            }

            it.btnBindService.setOnClickListener {
                // 绑定服务
                val bindIntent = Intent(this, BindService::class.java)
                bindService(bindIntent, connection, Service.BIND_AUTO_CREATE)
            }

            it.btnUnbindService.setOnClickListener {
                unbindService(connection)
            }

            it.btnGetCount.setOnClickListener {
                Toast.makeText(this, "${binder?.getCount()}", Toast.LENGTH_SHORT).show()
            }

            it.btnIntentService.setOnClickListener {
                MyIntentService.handleActionOne(this, "hello ", " IntentService")
            }
        }
    }
}